asterius/src/Asterius/Builtins/Primitive.hs

```diff
@@ -0,0 +1,187 @@
+-- @memset@, and @memcmp@ (see implementations in rts/rts.memory.mjs). Notice
+-- that for some of them we ignore their result type; the hsprimitive_*
+-- functions do not use them.
+primitiveImports :: [FunctionImport]
+primitiveImports =
+  [ mkImport "Memory" "memcpy" $
+      FunctionType {paramTypes = [F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memmove" $
+      FunctionType {paramTypes = [F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memset" $
+      FunctionType {paramTypes = [F64, F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memsetFloat" $
+      FunctionType {paramTypes = [F64, F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memcmp" $
+      FunctionType {paramTypes = [F64, F64, F64], returnTypes = [F64]}
+  ]
+
+-- -------------------------------------------------------------------------
+
+primitiveCBits :: AsteriusModule
+primitiveCBits =
+  primitiveMemcpy
+    <> primitiveMemmove
+    <> primitiveMemcmp
+    <> primitiveMemsetWord8
+    <> primitiveMemsetWord16
+    <> primitiveMemsetWord32
+    <> primitiveMemsetWord64
+    <> primitiveMemsetWord
+    <> primitiveMemsetPtr
+    <> primitiveMemsetFloat
+    <> primitiveMemsetDouble
+    <> primitiveMemsetChar
+
+-- -------------------------------------------------------------------------
+
+-- | @void hsprimitive_memcpy(void *dst, ptrdiff_t doff, void *src, ptrdiff_t soff, size_t len)@
+primitiveMemcpy :: AsteriusModule
+primitiveMemcpy = runEDSL "hsprimitive_memcpy" $ do
+  [dst, doff, src, soff, len] <- params [I64, I64, I64, I64, I64]
+  let arg1 = dst `addInt64` doff
+      arg2 = src `addInt64` soff
+  callImport "__asterius_Memory_memcpy" $
+    map convertUInt64ToFloat64 [arg1, arg2, len]
+
+-- | @void hsprimitive_memmove(void *dst, ptrdiff_t doff, void *src, ptrdiff_t soff, size_t len)@
+primitiveMemmove :: AsteriusModule
+primitiveMemmove = runEDSL "hsprimitive_memmove" $ do
+  [dst, doff, src, soff, len] <- params [I64, I64, I64, I64, I64]
+  let arg1 = dst `addInt64` doff
+      arg2 = src `addInt64` soff
+  callImport "__asterius_Memory_memmove" $
+    map convertUInt64ToFloat64 [arg1, arg2, len]
+
+-- | @int hsprimitive_memcmp(HsWord8 *s1, HsWord8 *s2, size_t n)@
+primitiveMemcmp :: AsteriusModule
+primitiveMemcmp = runEDSL "hsprimitive_memcmp" $ do
+  setReturnTypes [I64]
+  args <- params [I64, I64, I64]
+  truncSFloat64ToInt64
+    <$> callImport'
+      "__asterius_Memory_memcmp"
+      (map convertUInt64ToFloat64 args)
+      F64
+    >>= emit
+
+-- | @void hsprimitive_memset_XXX (XXX *p, ptrdiff_t off, size_t n, XXX x)@
+mkPrimitiveMemsetUInt ::
+  -- | Size (in bytes) of the type
+  Int ->
+  -- | String representation of the type
+  AsteriusEntitySymbol ->
+  AsteriusModule
+mkPrimitiveMemsetUInt size typerep = runEDSL hsname $ do
+  [p, off, n, x] <- params [I64, I64, I64, I64]
+  callImport "__asterius_Memory_memset" $
+    [ convertUInt64ToFloat64 $ p `addInt64` off,
+      convertSInt64ToFloat64 x,
```

Comment:
This should be `convertUInt64ToFloat64` as well. Since `x` is the element to be filled in the typed array, and all our typed arrays are created as unsigned versions

Reply:
I see. Does this apply to the `x` in `mkPrimitiveMemsetFloat` too?

Reply:
Ah, no it doesn't, I see it now. Ignore my comment